ark/jsonschema/object.ts

```diff
@@ -0,0 +1,336 @@
+			expected: "a valid object JSON Schema"
+		})
+	}
+
+	return {
+		optionalKeys: optionalKeys.map(key => ({
+			key,
+			value: innerParseJsonSchema.assert(jsonSchema.properties![key]).internal
+		})),
+		requiredKeys: requiredKeys.map(key => ({
+			key,
+			value: innerParseJsonSchema.assert(jsonSchema.properties![key]).internal
+		}))
+	}
+}
+
+const parseAdditionalProperties = (jsonSchema: JsonSchema.ObjectSchema) => {
+	if (!("additionalProperties" in jsonSchema)) return
+
+	const properties = Object.keys(jsonSchema.properties ?? {})
+	const patternProperties = Object.keys(jsonSchema.patternProperties ?? {})
+
+	const additionalPropertiesSchema = jsonSchema.additionalProperties
+	if (additionalPropertiesSchema === true) return
+
+	return (data: object, ctx: TraversalContext) => {
+		const errors: false[] = []
+
+		Object.keys(data).forEach(key => {
+			if (
+				properties.includes(key) ||
+				patternProperties.find(pattern => new RegExp(pattern).test(key))
+			)
+				// Not an additional property, so don't validate here
+				return
+
+			if (additionalPropertiesSchema === false) {
+				errors.push(
+					ctx.reject({
+						message: `property ${key} is an additional property, which the provided schema does not allow`
+					})
+				)
+				return
+			}
+
+			const additionalPropertyValidator = innerParseJsonSchema.assert(
+				additionalPropertiesSchema
+			)
+
+			const value = data[key as keyof typeof data]
+			if (!additionalPropertyValidator.allows(value)) {
+				errors.push(
+					ctx.reject({
+						problem: `property ${key} is an additional property so must adhere to additional property schema of ${additionalPropertyValidator.description} (was ${printable(value)})`
+					})
+				)
+			}
+		})
+		return errors.length === 0
+	}
+}
+
+export const validateJsonSchemaObject = JsonSchema.ObjectSchema.pipe(
+	(jsonSchema, ctx): Type<object> => {
+		const arktypeObjectSchema: Intersection.Schema<object> = {
+			domain: "object"
+		}
+
+		const { requiredKeys, optionalKeys } = parseRequiredAndOptionalKeys(
+			jsonSchema,
+			ctx
+		)
+		arktypeObjectSchema.required = requiredKeys
+		arktypeObjectSchema.optional = optionalKeys
+
+		const predicates: Predicate.Schema[] = [
+			...parseMinMaxProperties(jsonSchema, ctx),
+			parsePropertyNames(jsonSchema, ctx),
+			parsePatternProperties(jsonSchema, ctx),
+			parseAdditionalProperties(jsonSchema)
+		].filter(x => x !== undefined)
+
+		const typeWithoutPredicates = rootSchema(arktypeObjectSchema)
+		console.log(typeWithoutPredicates.json)
+		if (predicates.length === 0) return typeWithoutPredicates as never
+
+		return rootSchema({ domain: "object", predicate: predicates }).narrow(
+			(obj: object, innerCtx) => {
+				const validationResult = typeWithoutPredicates(obj)
+				if (validationResult instanceof ArkErrors) {
+					innerCtx.errors.merge(validationResult)
+					return false
+				}
+				return true
+			}
+		) as never
+	}
+)
+
+type inferAdditionalProperties<objectSchema> =
```

Comment:
Is there a reason we can't constrain this to a base object schema type so we don't have to keep casting key access?

Reply:
The issue is with the recursiveness.

When we do `inferAdditionalProperties<Omit<objectSchema, "aParsedProperty">>` to start parsing of the next property, it no longer meets that constraint.

E.g.:
![image](https://github.com/user-attachments/assets/6b7cd9eb-3535-4fdf-affd-2f2bbea48f2e)

Reply:
Marking as resolved since this is no longer relevant now that we've removed type inference (07a5215)